fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -1092,41 +1092,351 @@ inline bool tenantMapChanging(MutationRef const& mutation) {
+// INVALID_TENANT;
+inline int64_t lowerBoundTenantId(const StringRef& prefix, const std::map<int64_t, TenantName>& tenantMap) {
+	if (isSystemKey(prefix)) {
+		return TenantInfo::INVALID_TENANT;
+	}
+	Optional<int64_t> id;
+	if (prefix.size() >= TenantAPI::PREFIX_SIZE) {
+		// directly extract without copy
+		id = TenantAPI::extractTenantIdFromKeyRef(prefix);
+	} else {
+		// fall to generate a prefix
+		id = TenantIdCodec::lowerBound(prefix);
+	}
+	if (id.present() && id.get() >= 0) {
+		auto it = tenantMap.lower_bound(id.get());
+		if (it != tenantMap.end()) {
+			return it->first;
+		}
+	}
+	return TenantInfo::INVALID_TENANT;
+}
+
+TEST_CASE("/CommitProxy/SplitRange/LowerBoundTenantId") {
+	int mapSize = 1000;
+	std::map<int64_t, TenantName> tenantMap;
+	for (int i = 0; i < mapSize; ++i) {
+		tenantMap[i * 2] = ""_sr;
+	}
+
+	int64_t tid = lowerBoundTenantId(""_sr, tenantMap);
+	ASSERT_EQ(tid, 0);
+
+	tid = lowerBoundTenantId("\xff"_sr, tenantMap);
+	ASSERT_EQ(tid, TenantInfo::INVALID_TENANT);
+
+	tid = lowerBoundTenantId("\xff\x01\x02\x03\x04\x05\x06\x07\x08"_sr, tenantMap);
+	ASSERT_EQ(tid, TenantInfo::INVALID_TENANT);
+
+	tid = lowerBoundTenantId("\x99\x01\x02\x03\x04\x05\x06\x07\x08"_sr, tenantMap);
+	ASSERT_EQ(tid, TenantInfo::INVALID_TENANT);
+
+	int64_t targetId = deterministicRandom()->randomInt64(0, mapSize) * 2;
+	Key prefix = TenantAPI::idToPrefix(targetId);
+	tid = lowerBoundTenantId(prefix, tenantMap);
+	ASSERT_EQ(tid, targetId);
+
+	tid = lowerBoundTenantId(prefix.withSuffix("any"_sr), tenantMap);
+	ASSERT_EQ(tid, targetId);
+
+	targetId = deterministicRandom()->randomInt64(1, mapSize) * 2;
+	prefix = TenantAPI::idToPrefix(targetId - 1);
+	tid = lowerBoundTenantId(prefix, tenantMap);
+	ASSERT_EQ(tid, targetId);
+
+	targetId = deterministicRandom()->randomInt64(mapSize * 2, mapSize * 3);
+	prefix = TenantAPI::idToPrefix(targetId);
+	tid = lowerBoundTenantId(prefix, tenantMap);
+	ASSERT_EQ(tid, TenantInfo::INVALID_TENANT);
+
+	targetId = deterministicRandom()->randomInt64((int64_t)1 << 32, std::numeric_limits<int64_t>::max());
+	tenantMap[targetId] = ""_sr;
+	prefix = TenantAPI::idToPrefix(targetId);
+	int shift = deterministicRandom()->randomInt(0, TenantAPI::PREFIX_SIZE / 2);
+	prefix = prefix.substr(0, TenantAPI::PREFIX_SIZE - shift);
+	tid = lowerBoundTenantId(prefix, tenantMap);
+	ASSERT_EQ(tid, targetId);
+
+	return Void();
+}
+
+// Given a clear range [a, b), make a vector of clear range mutations split by tenant boundary [a, t0_end), [t1_begin,
+// t1_end), ... [tn_begin, b); The references are allocated on arena;
+std::vector<MutationRef> splitClearRangeByTenant(Arena& arena,
+                                                 const MutationRef& mutation,
+                                                 const std::map<int64_t, TenantName>& tenantMap) {
+	std::vector<MutationRef> results;
+	auto beginTenantId = lowerBoundTenantId(mutation.param1, tenantMap);
+	auto it = tenantMap.find(beginTenantId);
```

Comment:
If we're going to do a lookup of the tenant ID anyway, then probably we could get rid of the `lowerBoundTenantId` altogether and call `TenantIdCodec::lowerBound` here:

```
KeyRef prefix = mutation.param1.substr(0, std::min(mutation.param1.size(), 8));
Optional<int64_t> beginTenantId = TenantIdCodec::lowerBound(prefix);
auto it = beginTenantId.present() ? tenantMap.find(beginTenantId.get()) : tenantMap.end();
```

Reply:
Done, wrap the `find` in `lowerBoundTenantId` for unit test purpose